### PR TITLE
Improve error handling and WP-API integration

### PR DESCRIPTION
**Pull request by retrofox**
This PR proposes some changes not completely related each other.

### Error handling

We've started to make some requests using the WP-API. The structure responses aren't the same compared with the REST-API. An important thing that we have to handle is the response when the request fails.
First of all Let's clarify that we're using `envelope` as default for all requests. It means that we always get error `200` beyond of if the the request was failed or not.

**WP-API** -> http://v2.wp-api.org/reference/global.html#envelope

> ... some servers, clients, and proxies do not support accessing the full response data. The API supports passing an _envelope parameter, which sends all response data in the body, including headers and status code.

**REST-API**

> Some environments (like in-browser JavaScript or Flash) block or divert responses with a non-200 HTTP status code. Setting this parameter will force the HTTP status code to always be 200. The JSON response is wrapped in an "envelope" containing the "real" HTTP status code and headers.

When the request fails the REST-API responses data in the body similar to the following
image

And the next one is the response of WP-API:
image

So the goal with this PR is at least process the response and try to identify the error into the body for both APIs thus create an Error object to pass to the callback function. Something like 

```es6
import proxy from 'wpcom-proxy-request';

// requesting for `Me` using REST-API
proxy( {
  path: '/sites/en.blog.wordpress.com/users/me',
  apiVersion: '1.1'
}, function( error, body, headers ) {
  // ...
} );

// requesting for `Me` using WP-API
proxy( {
  path: '/sites/en.blog.wordpress.com/users/me',
  apiNamespace: 'wp/v2'
}, function( error, body, headers ) {
  // ...
} );
```

Now the error is not identified when the handler hits to WP-API.

### _headers

I've noticed that the _headers fields, injected into the body, isn't used. On the contrary we are deleting this field many times. So I think is a good idea not add it anymore and add a third parameter in the callback function which can be called `headers`.
Also, I've added a `status` field with the status code gotten from the body.

```json
{
  "Date": "Mon, 18 Jul 2016 21:00:29 GMT",
  "Content-Type": "application/json; charset=UTF-8",
  "status": 404
}
```

### Testing

I've created an application to run mocha tests in client-side. It simplify the authentication problem. The app can be run in the server with `make run-test-app` and we can _watch_ and _compile_ the code with `make watch-test-app`.

image

Open a tab pointing to `http://calypso.localhost:3001/` to see the tests results.

cc @timmyc @TooTallNate

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -410,9 +420,37 @@ function onmessage( e ) {
...
-	const body = data[ 0 ];
-	const statusCode = data[ 1 ];
-	const headers = data[ 2 ];
+	// Build `error` and `body` object from the `data` object
+	const { params } = xhr;
+	let body, statusCode, headers;
+
+	// detect `REST-API` of `WP-API` in function of `apiNamespace` parameter
+	debug( 'params: %o', params );
+	const isRestAPI = params.apiNamespace === undefined;
```
Comment (retrofox) [inline index.js]: Let's identify What API are we suing in function of `apiNamespace` value.

Comment (timmyc) [inline index.js]: maybe this should be `isWpAPI` since both are technically rest APIs

File: test/index.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+				}, ( error, data, headers ) => {
+					expect( error ).to.be.ok;
+					expect( data ).to.be.not.ok;
+					expect( headers ).to.be.ok;
+					expect( headers.status ).to.be.equal( 404 );
+					done();
+				} );
+			} );
+
+			it( 'should get `404` a non-existing site (_envelope: 1)', done => {
```
Comment (timmyc) [inline test/index.js]: should this read `non-existing post`?

File: index.js (hunk 1 above)
Comment (retrofox) [inline index.js]: :-D maybe `isRESTAPI`?

File: test/index.js (hunk 2 above)
Comment (retrofox) [inline test/index.js]: Right. Fixing and rebasing.

**Pull request merged**